packages/orama/src/trees/bkd.ts

```diff
@@ -0,0 +1,280 @@
+export interface Point {
+  x: number
+  y: number
+}
+
+export interface Node {
+  point: Point
+  left?: Node
+  right?: Node
+}
+
+type MaybeNode = Node | undefined
+
+interface InsertTask {
+  points: Point[]
+  depth: number
+  parent?: Node
+  direction?: 'left' | 'right'
+}
+
+interface SearchTask {
+  node: MaybeNode
+  depth: number
+}
+
+const K = 2 // 2D points
+const EARTH_RADIUS = 6371000 // meters
+
+export function create (points: Point[]): MaybeNode {
+  if (points.length === 0) {
+    return undefined
+  }
+
+  const stack: InsertTask[] = [{
+    points,
+    depth: 0
+  }]
+
+  let root: MaybeNode
+
+  while (stack.length > 0) {
+    const task = stack.pop()
+    if (task == null) {
+      continue
+    }
+
+    const { points, depth, parent, direction } = task
+
+    const axis = depth % K
+    points.sort((a, b) => (axis === 0 ? a.x - b.x : a.y - b.y))
+    const median = Math.floor(points.length / 2)
+    const node: Node = { point: points[median] }
+
+    if ((parent != null) && direction) {
+      parent[direction] = node
+    } else {
+      root = node
+    }
+
+    if (points.slice(0, median).length > 0) {
+      stack.push({
+        points: points.slice(0, median),
+        depth: depth + 1,
+        parent: node,
+        direction: 'left'
+      })
+    }
+
+    if (points.slice(median + 1).length > 0) {
```

Comment:
This is not the same as:

```suggestion
    if (median + 1 > (points.length - 1)) {
```